### ISSUE TITLE
Support user-defined ionic reversal potentials

### ISSUE THREAD

**Issue by halfflat** [arbor-sim/arbor#751]
Currently we just Nernst everything.

1. Determine and implement an API for describing desired reversal potential evaluation in the absence of a mechanism supplying the dynamics explicitly, i.e. constant or Nernst. This could be a good opportunity also for supplying other ion data — specifically initial interior or exterior concentrations — at a finer scale than the current global defaults allow.

2. Vectorized and GPU implementations for the Nernst updates that respect a subset of ion CV indices and incorporate weight data accordingly. These indices and weights will be determined by the settings from (1) above and the presence of any explicit dynamics.

**Comment by halfflat** [comment]
There is an additional issue: it's not safe to area-average multiple values for a reversal potential across the same control volume, unless the dynamics of that ion are the same across that c.v.

For example, if a control volume has weight α of one ionic current given by g₁(v-E₁), and 1-α of a current given by g₂(v-E₂), the weight averaged flux will be computed as (αg₁+(1-α)g₂)(v-(αE₁+(1-α)E₂)) instead of the correct flux of αg₁(v-E₁)+(1-α)g₂(v-E₂).

How to ameliorate?
1. We may adjust our control volume boundaries to match discontinuities in density mechanism dynamics.
2. We adjust the way that reversal potentials are described: the use of a common reversal potential across different ion channel models for the same species is essentially an optimization — the (for example) Nernst-derived potential is computed only once per control volume. Instead, we can require that models explicitly specify the form of the computed reversal potential they use to determine their ionic current contributions. Nernst-derived values can be provided by the library if required, avoiding some redundant computation, but other non-constant forms of the reversal potential would be specified within the ion current model as a function of membrane voltage, ion concentrations, and any other required parameters.

**Comment by halfflat** [comment]
Current on-going work goes with (2), more or less. WIP, but reversal potential calculations can be specified for an ion at the cellular level, but not more finely, avoiding the potential nonlinearities described above.

For generic Nerst equation implementations, we need a way of feeding in ionic valence through to a mechanism (among other things). This was becoming a huger ball of changes, so moving that into another issue.

**Comment by halfflat** [comment]
Ionic valence reading in mechanisms was made as #780, implemented in #781. Will make another issue and PR for re-parameterizing mechanisms over their ion channels, so that we can have, for example, a single generic Nernst mechanism that is specialized by the user for any ion they wish.
